R/redist_smc.R

```diff
@@ -278,6 +380,79 @@ redist_smc = function(map, nsims, counties=NULL, compactness=1, constraints=list
 #'
 redist_quantile_trunc = function(x) pmin(x, quantile(x, 1 - length(x)^(-0.5)))
 
+
+#' Confidence Intervals for SMC Estimates
+#'
+#' Builds a confidence interval for a quantity of interest.
+#' If multiple runs are available, uses the between-run variation to estimate
+#' the standard error. If only one run is available, uses information on the SMC
+#' particle/plan genealogy to estimate the standard error, using a variant of
+#' the method of Olson & Douc (2019). The multiple-run estimator is more
+#' reliable, especially for situations with many districts, and should be used
+#' when parallelism is available.  All reference plans are ignored.
+#'
+#' @param plans a [redist_plans] object.
+#' @param x the quantity to build an interval for. Tidy-evaluated within `plans`.
+#' @param district for [redist_plans] objects with multiple districts, which
+#'   `district` to subset to. Set to `NULL` to perform no subsetting.
+#' @param conf the desired confidence level.
+#'
+#' @references
+#' Lee, A., & Whiteley, N. (2018). Variance estimation in the particle filter.
+#' Biometrika, 105(3), 609-625.
+#' Olsson, J., & Douc, R. (2019). Numerically stable online estimation of
+#' variance in particle filters. Bernoulli, 25(2), 1504-1535.
+#' H. P. Chan and T. L. Lai. A general theory of particle filters in hidden
+#' Markov models and some applications. Ann. Statist., 41(6):2877–2904, 2013.
+#'
+#' @return a vector of length 3: (lower, point estimate, upper).
+#'
+#' @examples
+#' library(dplyr)
+#' data(iowa)
+#'
+#' iowa_map = redist_map(iowa, existing_plan=cd_2010, pop_tol=0.05)
+#' plans = redist_mergesplit_parallel(iowa_map, nsims=200, chains=2, silent=TRUE) %>%
+#'     mutate(dem = group_frac(iowa_map, dem_08, dem_08 + rep_08)) %>%
+#'     number_by(dem)
+#' redist_smc_ci(plans, dem)
+#'
+#' @md
+#' @concept analyze
+#' @export
+redist_smc_ci = function(plans, x, district=1L, conf=0.9) {
```

Comment:
Is default 0.9 best?

Reply:
I think so. 95% aren't very stable. 5th to 95th percentile seems good